dnf/cli/commands/history.py

```diff
@@ -0,0 +1,506 @@
+                        else:
+                            logger.info(msg)
+
+        return sorted(tids, reverse=True), merged_tids
+
+    def _verify_toplevel_json(self):
+        fn = self.opts.transaction_filename
+
+        if "version" not in self.replay_data:
+            raise TransactionFileError(fn, _('Missing key "{key}".'.format(key="version")))
+
+        if not isinstance(self.replay_data["version"], str):
+            raise TransactionFileError(fn, _('Unexpected type of "version", string expected.'))
+
+        check_version(self.replay_data["version"], fn)
+
+    def _replay_pkg_action(self, na, action_data):
+        try:
+            na_str = "%s.%s" % (na.name, na.arch)
+            action = action_data["action"]
+
+            # validate the reasons before running the transaction and cache them
+            try:
+                self._na_reason_cache[na_str] = \
+                    libdnf.transaction.StringToTransactionItemReason(action_data["reason"])
+            except IndexError as e:
+                raise TransactionError(
+                    _('Unexpected value of package reason "{reason}" for package "{pkg}".')
+                        .format(reason=action_data["reason"], pkg=na_str)
+                )
+
+            if action in ("Install", "Upgrade", "Downgrade", "Reinstall"):
+                nevra = "%s-%s.%s" % (na.name, action_data["to"], na.arch)
+                if action in ("Upgrade", "Downgrade", "Reinstall"):
+                    from_nevra = "%s-%s.%s" % (na.name, action_data["from"], na.arch)
+                    query = dnf.subject.Subject(from_nevra).get_best_solution(
+                        self.base.sack,
+                        with_nevra=True,
+                        with_provides=False,
+                        with_filenames=False,
+                        forms=[hawkey.FORM_NEVRA]
+                    )["query"]
+
+                    if not query.installed():
+                        raise TransactionError(
+                            _('Package "{pkg}" not installed for action "{action}".')
+                                .format(pkg=from_nevra, action=action)
+                        )
+
+            elif action == "Removed" or action == "Obsoleted":
+                nevra = "%s-%s.%s" % (na.name, action_data["from"], na.arch)
+            else:
+                raise TransactionError(
+                    _('Unexpected value of package action "{action}" for package "{pkg}".')
+                        .format(action=action, pkg=na_str)
+                )
+
+            query = dnf.subject.Subject(nevra).get_best_solution(
+                self.base.sack,
+                with_nevra=True,
+                with_provides=False,
+                with_filenames=False,
+                forms=[hawkey.FORM_NEVRA]
+            )["query"]
+
+            if not query:
+                raise TransactionError(_('Package "{nevra}" was not found.').format(nevra=nevra))
```

Comment:
What about to be more flexible with --skip-broken option. This option is very important for any history transaction command because otherwise it will be impossible to use them on Fedora.